# **Telecom X - Análisis de Evasión de Clientes**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import json

##📌 **Extracción**

In [3]:
# Cargar el archivo JSON
with open('/content/drive/MyDrive/Alura-Challenge-Telecom/TelecomX_Data.json', 'r') as file:
    data = json.load(file)


In [5]:
# Convertir los datos a un DataFrame de pandas
df = pd.DataFrame(data)
df

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7267 rows x 6 columns]

In [8]:
# Convertir estructuras anidadas en DataFrame plano
df_raw = pd.json_normalize(data)
df_raw

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [9]:
df_raw.shape

(7267, 21)

## 🔧 **Transformación**

In [10]:
# Filtrar registros válidos (Churn 'Yes' o 'No')
df_clean = df_raw[df_raw['Churn'].isin(['Yes', 'No'])].copy()

In [11]:
# Convertir 'Total Charges' a numérico
df_clean['account.Charges.Total'] = pd.to_numeric(df_clean['account.Charges.Total'], errors='coerce')

In [12]:
# Eliminar filas con valores nulos en 'Total Charges'
df_clean = df_clean.dropna(subset=['account.Charges.Total'])

In [13]:
# Revisar estructura final
print("Dimensiones después de limpieza:", df_clean.shape)
print(df_clean.dtypes)

Dimensiones después de limpieza: (7032, 21)
customerID                    object
Churn                         object
customer.gender               object
customer.SeniorCitizen         int64
customer.Partner              object
customer.Dependents           object
customer.tenure                int64
phone.PhoneService            object
phone.MultipleLines           object
internet.InternetService      object
internet.OnlineSecurity       object
internet.OnlineBackup         object
internet.DeviceProtection     object
internet.TechSupport          object
internet.StreamingTV          object
internet.StreamingMovies      object
account.Contract              object
account.PaperlessBilling      object
account.PaymentMethod         object
account.Charges.Monthly      float64
account.Charges.Total        float64
dtype: object


## **📊 Carga y análisis**

In [14]:
# 1. Tasa general de churn
churn_rate = df_clean['Churn'].value_counts(normalize=True) * 100
print("Tasa general de churn (%):\n", churn_rate.round(2))

Tasa general de churn (%):
 Churn
No     73.42
Yes    26.58
Name: proportion, dtype: float64


In [15]:
# 2. Churn por tipo de contrato
churn_by_contract = df_clean.groupby(['account.Contract', 'Churn']).size().unstack().fillna(0)
churn_by_contract_percent = churn_by_contract.div(churn_by_contract.sum(axis=1), axis=0) * 100
print("\nChurn por tipo de contrato (%):\n", churn_by_contract_percent.round(2))


Churn por tipo de contrato (%):
 Churn                No    Yes
account.Contract              
Month-to-month    57.29  42.71
One year          88.72  11.28
Two year          97.15   2.85


In [16]:
# 3. Churn por tipo de Internet
churn_by_internet = df_clean.groupby(['internet.InternetService', 'Churn']).size().unstack().fillna(0)
churn_by_internet_percent = churn_by_internet.div(churn_by_internet.sum(axis=1), axis=0) * 100
print("\nChurn por tipo de Internet (%):\n", churn_by_internet_percent.round(2))


Churn por tipo de Internet (%):
 Churn                        No    Yes
internet.InternetService              
DSL                       81.00  19.00
Fiber optic               58.11  41.89
No                        92.57   7.43


In [17]:
# 4. Churn por método de pago
churn_by_payment = df_clean.groupby(['account.PaymentMethod', 'Churn']).size().unstack().fillna(0)
churn_by_payment_percent = churn_by_payment.div(churn_by_payment.sum(axis=1), axis=0) * 100
print("\nChurn por método de pago (%):\n", churn_by_payment_percent.round(2))


Churn por método de pago (%):
 Churn                         No    Yes
account.PaymentMethod                  
Bank transfer (automatic)  83.27  16.73
Credit card (automatic)    84.75  15.25
Electronic check           54.71  45.29
Mailed check               80.80  19.20


#📄**Informe final**

## **Factores de Abandono en Clientes de TelecomX**

**1. Tasa General de Abandono**

- El 26.58% de los clientes han abandonado la compañía.

- El 73.42% de los clientes se mantienen activos.

📌**Esto representa una tasa alta de churn para una empresa de telecomunicaciones, lo cual requiere atención inmediata.**

**2. Relación entre Tipo de Contrato y Churn**

**a) Tipo de Contrato Mes a mes:** Churn(%) 42.71 / Retención(%) 57.29

**b) Tipo de Contrato 1 año:**	   Churn(%) 11.28 / Retención(%) 88.72

**c) Tipo de Contrato 2 años:**    Churn(%)  2.85 / Retención(%) 97.15

📌**Los contratos de largo plazo (1 y 2 años) retienen mucho mejor. Los clientes con contratos mes a mes son mucho más propensos a cancelar.**



**3. Relación entre Tipo de Internet y Churn**

**a) Fibra Óptica:** Churn(%) 41.89 / Retención(%) 58.11

**b) DSL:** Churn(%) 19.00 / Retención(%) 81:00

**c) Sin Servicio:** Churn(%) 7.43 / Retención(%) 92.57

📌**Los usuarios con fibra óptica presentan la mayor tasa de churn, lo que podría indicar una brecha entre la promesa y la experiencia del servicio.**




**4. Relación entre Método de Pago y Churn**

**a) Cheque Electrónico:** Churn(%) 45.29 / Retención(%) 54.71

**b) Cheque por Correo:** Churn(%) 19.20 / Retención(%) 80:80

**c) Transferencia Automática:** Churn(%) 16.73 / Retención(%) 83.27

**d) Tarjeta Automática:** Churn(%) 15.25 / Retención(%) 84.75

📌**El método de pago más riesgoso es el cheque electrónico. Los pagos automáticos están asociados con mayor retención.**



✅**Recomendaciones Iniciales

**1. Fomentar contratos largos mediante descuentos o beneficios exclusivos.**

**2. Auditar el servicio de fibra óptica, especialmente la calidad percibida por el cliente.**

**3. Incentivar métodos de pago automáticos, como transferencias o tarjetas, con promociones.**

